py/persipubsub/queue.py

```diff
@@ -0,0 +1,489 @@
+                 max_msgs_num: Optional[int] = None,
+                 hwm_lmdb_size_bytes: Optional[int] = None) -> None:
+        """
+        Initialize.
+
+        :param msg_timeout_secs: time after which msg is classified as dangling
+        msg (secs)
+        :param max_msgs_num: maximal amount of msg
+        :param hwm_lmdb_size_bytes: high water mark for total size of LMDB
+        (bytes)
+        """
+        self.msg_timeout_secs = MSG_TIMEOUT_SECS \
+            if msg_timeout_secs is None else msg_timeout_secs  # type: int
+        self.max_msgs_num = MAX_MSGS_NUM \
+            if max_msgs_num is None else max_msgs_num  # type: int
+        self.hwm_lmdb_size_bytes = HWM_LMDB_SIZE_BYTES \
+            if hwm_lmdb_size_bytes is None else hwm_lmdb_size_bytes  # type: int
+
+
+def _initialize_environment(
+        queue_dir: pathlib.Path,
+        max_reader_num: int = 1024,
+        max_db_num: int = 1024,
+        max_db_size_bytes: int = 32 * 1024**3) -> lmdb.Environment:
+    """
+    Initialize the queue; the queue directory is assumed to exist.
+
+    :param queue_dir: where the queue is stored
+    :param max_reader_num: maximal number of readers
+    :param max_db_num: maximal number of databases
+    :param max_db_size_bytes: maximal size of database (bytes)
+    :return: Load or if needed create LMDB from directory
+    """
+    if not queue_dir.exists():
+        raise RuntimeError(
+            "The queue directory does not exist: {}".format(queue_dir))
+
+    env = lmdb.open(
+        path=queue_dir.as_posix(),
+        map_size=max_db_size_bytes,
+        subdir=True,
+        max_readers=max_reader_num,
+        max_dbs=max_db_num)
+    return env
+
+
+def _prune_dangling_messages_for(queue: '_Queue',
+                                 subscriber_ids: List[str]) -> None:
+    """
+    Prune all dangling messages for subscribers of a queue from LMDB.
+
+    :param queue: of which dangling messages should be pruned
+    :param subscriber_ids: subscribers of which dangling msgs should be pruned
+    """
+    assert isinstance(queue.env, lmdb.Environment)
+    with queue.env.begin(write=True) as txn:
+        pending_db = queue.env.open_db(
+            key=persipubsub.PENDING_DB, txn=txn, create=False)
+        meta_db = queue.env.open_db(
+            key=persipubsub.META_DB, txn=txn, create=False)
+        data_db = queue.env.open_db(
+            key=persipubsub.DATA_DB, txn=txn, create=False)
+
+    # Definition of dangling messages:
+    #   - having no pending subscribers
+    #   - exists longer than timeout allows
+    msgs_to_delete = set()
+    with queue.env.begin(db=pending_db) as txn:
+        cursor = txn.cursor()
+        cursor.first()
+        for key, pending_subscribers_num in cursor:
+            if persipubsub.bytes_to_int(pending_subscribers_num) == 0:
+                msgs_to_delete.add(key)
+
+    # subscriber might still await these messages after the timeout.
```

Comment:
Please explain this aspect in more detail. It is not clear to me why the subscribers might wait for these messages after the timeout since I assume everything takes place in a single transaction?